Closes #2444 - SegArray with String Values Parquet Support

Ethan-DeBandi99 opened the pull request:
Closes #2444 

This PR adds support for writing SegArray objects with Strings values to Parquet. 

**This does not add support for the multi-column case. That will be handled separately.**

pierce314159 commented on src/ParquetMsg.chpl:
```diff
@@ -1049,7 +1038,88 @@ module ParquetMsg {
+      const myFilename = filenames[idx];
+
+      const locDom = segments.localSubdomain();
+      var dims: [0..#1] int;
+      dims[0] = locDom.size: int;
```
it doesn't seem like `dims` is used

Ethan-DeBandi99 replied:
You are right. Removed.

pierce314159 reviewed (approved):
Looks good to me! Nice job on this, it seems like this required _ a lot_ of attention to detail to get the localization in `writeStrSegArrayParquet` just right

pierce314159 reviewed (changes requested):
oops sorry I approved to soon. I just pulled down and to verify correctness with gasnet and im seeing an error when running the test with 4 locales

```python
ARKOUDA_RUNNING_MODE=CLIENT python3 -m pytest tests/parquet_test.py -k "test_segarray_string"
E           AssertionError: Lists differ: ['one', 'two', 'three', 'uno', 'dos', 'tres'] != ['one', 'two', 'ne\x00tw', 'uno', 'dos', 'no\x00d']
E
E           First differing element 2:
E           'three'
E           'ne\x00tw'
E
E           - ['one', 'two', 'three', 'uno', 'dos', 'tres']
E           ?                  ^^^^                  ^^^^
E
E           + ['one', 'two', 'ne\x00tw', 'uno', 'dos', 'no\x00d']
E           ?                 ++++++ ^                  ^^^^^^^
```

pierce314159 commented:
With 3 locales, im actually seeing this
```
 *** Incorrect guard value: 92292502241651
arkouda_server_real(9547,0x16fb24000) malloc: *** set a breakpoint in malloc_error_break to debug
*** Caught a fatal signal (proc 1): SIGABRT(6)
```
just spoke with ethan and we agreed the right way forward is to close this and reopen when these have been addressed

Ethan-DeBandi99 commented on src/ArrowFunctions.cpp:
```diff
@@ -1251,7 +1304,91 @@ int cpp_writeStrColumnToParquet(const char* filename, void* chpl_arr, void* chpl
+        parquet::ByteArrayWriter* ba_writer =
+          static_cast<parquet::ByteArrayWriter*>(rg_writer->NextColumn());
+        int64_t count = 0;
+        while (numLeft > 0 && count < rowGroupSize) { // ensures rowGroupSize maintained
+          int64_t segmentLength = segments[segIdx+1] - segments[segIdx];
```
We will not because segments is artificially padded with the total # of values as the last element.

Ethan-DeBandi99 commented on src/ArrowFunctions.cpp:
```diff
@@ -1251,7 +1304,91 @@ int cpp_writeStrColumnToParquet(const char* filename, void* chpl_arr, void* chpl
+            int16_t def_lvl = 3;
+            parquet::ByteArray value;
+            value.ptr = reinterpret_cast<const uint8_t*>(&chpl_ptr[valIdx]);
+            value.len = offsets[offIdx+1] - offsets[offIdx] - 1;
+            ba_writer->WriteBatch(1, &def_lvl, &rep_lvl, &value);
+            offIdx++;
+            valIdx+=offsets[offIdx] - offsets[offIdx-1];
```
I don't remember exactly what the issue was here, but there was a specific reason I did it this way. I can dig into it more if you would like.

Ethan-DeBandi99 commented:
@pierce314159 - Everything is functioning properly now. Thank you for checking initially. I had a configuration issue in my environment that caused me to miss the failures. I verified that everything is good to go now. I also addressed your comments from elsewhere in the code.

pierce314159 reviewed (approved):
Looks good! I verified the test passes with a couple different number of locales

jaketrookman reviewed (approved):
Looks good
